### ISSUE TITLE
Error when executing test.py on a smaller subset of MSCOCO2014

### ISSUE THREAD

**Issue by kushv16** [jchenghu/ExpansionNet_v2#27]
Hi,

I am trying to use the model weights you provided on the google drive to replicate results on a smaller set of MSCOCO2014 data.

I have a different annotations.json file which has around 14000 images from 2014 MS coco. 

This is the command I am trying to execute on google colab:

!python test.py --N_enc 3 --N_dec 3 --model_dim 512 \
    --num_gpus 1 --eval_beam_sizes [5] --is_end_to_end True \
    --eval_parallel_batch_size 4 \
    --captions_path /content/filtered_data.json \
    --images_path /content/data_hw4/dataHm4/data/ \
    --save_model_path /content/drive/MyDrive/rf_model_140.4.pth

The error that I am seeing is attached below as an image. 

My end goal is to use your model weights to evaluate the model on a smaller subset of data. 

![image](https://github.com/user-attachments/assets/46777168-c779-442d-84c2-0b7cf4b493ed)

**Comment by jchenghu** [comment]
Hi, sorry for the late response, my workstation broke recently and I m currently posting this answer with the phone, sadly, I might not be able to provide a proper code..

That being said, the problem is that, the vocabulary created from your subset of coco, contains only 3870 tokens, instead of 10000 of the currently pretrained model.

Actually im a little bit confused about the goal, you said you want to evaluate the model on a smaller subset of data, made of 14.000 images but the test.py performs evaluation over the validation and test sets, both made of 5000 each, so I think your data is bigger at least for the testing phase. 

Because of that, I have to assume that the image id of your 14k images can be in the training, testing and validation set arbitrarily....  which might complicate the solution..

So, I m assuming you want to perform evaluation over your custom json with 14.000 images only from coco, I think a possible solution would be this:

Inside the 'data/coco_dataset.py', write

    TrainSet_ID = 1
    ValidationSet_ID = 2
    TestSet_ID = 3
    MY_CUSTOM_SET_ID = 4  <-- this

then

        self.karpathy_train_list = []
        self.karpathy_val_list = []
        self.karpathy_test_list = []
        for key in self.karpathy_train_dict.keys():
            self.karpathy_train_list.append(self.karpathy_train_dict[key])
        for key in self.karpathy_val_dict.keys():
            self.karpathy_val_list.append(self.karpathy_val_dict[key])
        for key in self.karpathy_test_dict.keys():
            self.karpathy_test_list.append(self.karpathy_test_dict[key])
       # here, load your cutom list of id and captions 
      with open(ur json path) as f:
          my_json = json.load(f)
          do processing of ur json to populate the 
             'self.my_custom_list' such that it is a list of dictionaries formatted as
          self.my_custom_list = [
              {'image_id': integer id, 'captions': list captions, 'image_path': path to the images which is id and and padd, # hint: u can create it in this way 'img_path': self.images_path + json_item['filepath'] + '/img/' + json_item['filename']
               }, { instance 2 ...}, ..., { instance 14000} 
                
          ]
     self.num_custom_set = len(self.my_custom_list)


Now, go to coco_dataloader.py, and add a few lines here

 ```
def get_images_by_idx(self, img_idx, dataset_split, transf_mode='train', get_also_id=False):
        if dataset_split == CocoDatasetKarpathy.TestSet_ID:
            file_path = self.coco_dataset.karpathy_test_list[img_idx]['img_path']
            img_id = self.coco_dataset.karpathy_test_list[img_idx]['img_id']
        elif dataset_split == CocoDatasetKarpathy.ValidationSet_ID:
            file_path = self.coco_dataset.karpathy_val_list[img_idx]['img_path']
            img_id = self.coco_dataset.karpathy_val_list[img_idx]['img_id']

# ... here add these lines
dataset_split == CocoDatasetKarpathy.MY_CUSTOM_SET_ID:
            file_path = self.coco_dataset.my_custom_list[img_idx]['img_path']
            img_id = self.coco_dataset.my_custom_list[img_idx]['img_id']

        else:
            file_path = self.coco_dataset.karpathy_train_list[img_idx]['img_path']
            img_id = self.coco_dataset.karpathy_train_list[img_idx]['img_id']

 ``` 

Finally, in the test.py, add this line after or before the invokation on validation and test

```

 print("Evaluation on Custom Set")
    evaluate_model_on_set(ddp_model, coco_dataset.caption_idx2word_list,
                          coco_dataset.get_sos_token_idx(), coco_dataset.get_eos_token_idx(),
                          coco_dataset.num_custom_set,
                          data_loader,
                          CocoDatasetKarpathy.MY_CUSTOM_SET_ID, model_max_len,
                          rank, ddp_sync_port,
                          parallel_batches=eval_parallel_batch_size,
                          use_images_instead_of_features=is_end_to_end,
                          beam_sizes=eval_beam_sizes)

```
Now, it is important to launch test.py with my dataset.json, not yours, to preserve the token mapping of the vocabulary. Your json will be loaded inside the dataset creation

The solution is not elegant, but it should work and I believe it is the most straightforward

Let me know if it helped

**Comment by kushv16** [comment]
Hi,

Thanks for getting back.

Let me explain you my use case in more detail. 

I have a smaller MSCOCO2014 dataset with 14000 items. 13004 of these belong to "split" == train or "split" == restval (So this becomes my training set). 490 images belong to split "val" and 506 images belong to split "test". 

So, I want to do two things — 

part-1: Train the entire pipeline on a dataset of total 14000 items. I was successful with this, I was able to use your code to train all the stages explained in your readme. 

part-2: Use the model weights you provided in the google drive to perform testing on those 490 (val) 506 (test) images that have split as "val" and "test" respectively. 

<img width="996" alt="image" src="https://github.com/user-attachments/assets/164cf2e5-f48a-4eb5-a189-13768822491e">

These are my results from part-1 on those 490 (val) and 506 (test) images.

Now, for the second part, I want to use your model weights to see what the scores look like on the same 490 (val) and 506 (test) images.

I need help with the second part, where I am using your model weights to run the following command: 

```
!python test.py --N_enc 3 --N_dec 3 --model_dim 512 \
    --num_gpus 1 --eval_beam_sizes [5] --is_end_to_end False \
    --eval_parallel_batch_size 4 \
    --images_path /content/data/ \
    --save_model_path /content/ExpansionNet_v2/github_ignore_material/saves/checkpoint_2024-12-04-21:50:24_epoch8it541bs24_rf_.pth \
    --captions_path /content/annotation.json \
    --features_path /content/ExpansionNet_v2/github_ignore_material/raw_data/features.hdf5
```

This is one entry from my annotation.json, hoping you get more clarity by looking at this JSON. 

```
 {
            {
            "filepath": "val2014",
            "sentids": [
                501985,
                507367,
                508369,
                510313,
                510664
            ],
            "filename": "COCO_val2014_000000403385.jpg",
            "imgid": 39,
            "split": "test",
            "sentences": [
                {
                    "tokens": [
                        "a",
                        "bathroom",
                        "that",
                        "has",
                        "a",
                        "broken",
                        "wall",
                        "in",
                        "the",
                        "shower"
                    ],
                    "raw": "A bathroom that has a broken wall in the shower.",
                    "imgid": 39,
                    "sentid": 501985
                },
                {
                    "tokens": [
                        "a",
                        "bathroom",
                        "looks",
                        "clean",
                        "but",
                        "is",
                        "missing",
                        "tile",
                        "at",
                        "the",
                        "shower",
                        "stall"
                    ],
                    "raw": "A bathroom looks clean but is missing tile at the shower stall.",
                    "imgid": 39,
                    "sentid": 507367
                },
                {
                    "tokens": [
                        "a",
                        "view",
                        "of",
                        "a",
                        "bathroom",
                        "that",
                        "needs",
                        "to",
                        "be",
                        "fixed",
                        "up"
                    ],
                    "raw": "A view of a bathroom that needs to be fixed up.",
                    "imgid": 39,
                    "sentid": 508369
                },
                {
                    "tokens": [
                        "a",
                        "shower",
                        "toilet",
                        "and",
                        "sink",
                        "in",
                        "a",
                        "basement",
                        "bathroom"
                    ],
                    "raw": "a shower toilet and sink in a basement bathroom",
                    "imgid": 39,
                    "sentid": 510313
                },
                {
                    "tokens": [
                        "a",
                        "very",
                        "big",
                        "whit",
                        "rest",
                        "room",
                        "with",
                        "a",
                        "shabby",
                        "looking",
                        "shower"
                    ],
                    "raw": "A very big whit rest room with a shabby looking shower.",
                    "imgid": 39,
                    "sentid": 510664
                }
            ],
            "cocoid": 403385
        },
```

**Comment by jchenghu** [comment]
Hi,

Thank you for the detailed explanation, it is much clearer now :-)

I think the most straightforward solution would be:
- Load your custom dataset
- Load the original vocabulary of `rf_model_140.4.pth`

To do that, let's stick with your first command:
```
!python test.py --N_enc 3 --N_dec 3 --model_dim 512
--num_gpus 1 --eval_beam_sizes [5] --is_end_to_end True
--eval_parallel_batch_size 4
--captions_path /content/filtered_data.json
--images_path /content/data_hw4/dataHm4/data/
--save_model_path /content/drive/MyDrive/rf_model_140.4.pth
```
I'm assuming that this loads your training (14k), test (506), and validation (490) sets

Let's now add a few lines inside `data/coco_dataset.py`, locate these lines:
```
        discovered_words.sort()
        self.caption_word2idx_dict = dict()
        self.caption_idx2word_list = []
        for i in range(len(discovered_words)):
            self.caption_word2idx_dict[discovered_words[i]] = i
            self.caption_idx2word_list.append(discovered_words[i])
        if verbose:
            print("There are " + str(self.num_caption_vocab) + " vocabs in dict")

       # ---------------------------------------------------------------------------------------
       # Add these additional lines here...
       with open('./demo_material/demo_coco_tokens.pickle', 'rb') as f:
           coco_tokens = pickle.load(f)
           self.caption_idx2word_list = coco_tokens['idx2word_list']
           self.caption_word2idx_dict = coco_tokens['word2idx_dict']
```
This takes care of the difference in the vocabulary size and how each token is mapped in the model vocabulary projection. The data loader will address the conversion from words to tokens automatically, no more operations should be needed.

Let me know if it works!

Best regards,
Jia

**Comment by kushv16** [comment]
Hi Jia,

I followed your steps and updated my data/coco_dataset.py

This is the new error that I am getting: 

![image](https://github.com/user-attachments/assets/de5d75aa-b69e-434b-a1c4-074ca2655d41)

**Comment by jchenghu** [comment]
Hi,
You can fix that by manually setting the maximum sequence length observed in the that data set,
add `self.max_seq_len = 74 ` here
```

        discovered_words.sort()
        self.caption_word2idx_dict = dict()
        self.caption_idx2word_list = []
        for i in range(len(discovered_words)):
            self.caption_word2idx_dict[discovered_words[i]] = i
            self.caption_idx2word_list.append(discovered_words[i])
        if verbose:
            print("There are " + str(self.num_caption_vocab) + " vocabs in dict")

       # ---------------------------------------------------------------------------------------
       # Add these additional lines here...
       with open('./demo_material/demo_coco_tokens.pickle', 'rb') as f:
           coco_tokens = pickle.load(f)
           self.caption_idx2word_list = coco_tokens['idx2word_list']
           self.caption_word2idx_dict = coco_tokens['word2idx_dict']
       self.max_seq_len = 74 # <-- There
```

**Comment by kushv16** [comment]
Thanks for the help!
